README
This is rbldnsd.

The current source tree is mantained by The Spamhaus Project Ltd.
The original source was written originally by Michael Tokarev <mjt+rbldnsd@corpit.ru>


The current source and documentation can be found at:
  https://rbldnsd.io/


The original, unmantained source and debian packages can always be found at:
  http://www.corpit.ru/mjt/rbldnsd.html


Copyright (C) 2016-2018 The Spamhaus Project Ltd.
Copyright (C) 2002 Michael Tokarev

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along
  with this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


NEWS
This file describes user-visible changes in rbldnsd.
Newer news is at the top.

1.0pre (Still not official, to be released)
 - Empty Non Terminals patch. This is a compile-time option and
   is meant to address some incompatibilities with RFC 7816.
   Adding the "$ENT" special entity to all the datasets.
 - Support for systemd added (thanks to Marco d'Itri). Please
   use --enable-systemd when configuring and make sure that you have
   pkg-config installed.
 - Test suite migrated over to python-3. Thanks to
   Michael Orlitzky for the contributions.
 - Fix a longstanding bug in the base-template replacement
 - Removal of deprecated features (aka: NS record compatibility mode)
 - Adding -F flag, used to identify the log facility of the daemon.
 - fix tests for systems without ipv6 support, or when ipv6 is
   disabled in rbldnsd at compile-time
 - fix tests for API change in pydns >= 2.3.6
 - It is no longer an error to request binding to a particular
   address/port more than once. (The subsequent requests are simply
   ignored.) (This avoids confusion on certain systems/configurations
   where gethostbyname("localhost") can return 127.0.0.1 multiple
   times.)
 - Moving rbldnsd.spec file under contrib/rpm/ to match debian move
 - Updating various pointers to point to new contrib/ locations
   (both debian and rpm spec)
 - Maintenance of RPM spec file to match new file locations, versions,
   and the make dist tar file naming schema.
 - Fix for sockaddr_equal() which was missing an IPv6 case.

0.998b (21 Dec 2016)
 - Fix for memory errors on very large datasets.
   Patch by Andrew Clayton

0.998a (06 Sep 2016)
 - Minor fixes in copyright and documentation
 - bugfix: minor fix to prevent errors on newer compilers
 - Moving debian/ old directory under contrib/debian/

0.998 (05 Dec 2015)

 - bugfix: correctly handle V4MAPPED (v4 in v6) addresses, the
   original v6 prefix was wrong.  Thanks to Alex Lasoriti for
   finding the issue
 - bugfix: sometimes IP4-based datasets gave false positives when
   an IP6 dataset were present, and it was also possible to have
   false positive in IP6 datasets.  Both has been fixed.

0.997a (23 Jul 2013)

 - minor fixes/changes in packaging, no code changes.
   In particular, fixes a build failure on *BSD introduced in 0.997

0.997 (29 Jun 2013)

 - main feature of this version is ipv6 support.  Many thanks to
   Geoffrey T. Dairiki for the implementation of btrie (btrie.c) which
   is far superior to old ip4trie code and handles both v4 and v6

 - feature: ip6trie - new dataset supports listing of arbitrary length
   ip6 CIDRs, along with individual A/TXT values for each prefix

 - feature: ip6tset - new dataset supports listing of ip6 /64 subnets
   and the exclusion of /128 subnets; only supports a single A/TXT value
   for the entire dataset

 - optimization: ip4trie - using new trie implementation (developed for
   the ip6trie dataset) decreases memory consumption by roughly a factor
   of three

 - feature: acl dataset - ip6 addresses are now supported in ACLs

 - feature: added --enable-asserts configure option to enable
   compilation of debugging assertions; assertion checking is
   disabled by default

 - featurette: zero-length "wildcard" IP4 CIDR prefixes are now
   allowed in ip4trie and acl datasets.

0.996b (29 Mar 2008)

 - most stuff is very minor, while preparing for larger changes
   for 0.997 and 1.0.

 - several (mostly minor) bug and portability fixes.

 - feature: ability to specify "base template" for a dataset, using $=
   variable.  See manpage for details, "Resulting A values and TXT templates"
   section.

 - incompat: due to the above change ($= base template), in TXT entries which
   starts with equal sign (=), the first character (which is `=', obviously)
   is removed.  $= is now treated specially too.

 - feature: (experimental) support for dynamically-loaded extensions (DSOs),
   in place of (previously unfinished) compile-time "hooks" support.
   New options -x/-X, to load an extension and to specify argument for it.
   Usage document has to be written still.

0.996a (27 Jul 2006) (The "34-Birthday Release")

 - -a/-A (-A is new, currently a noop) option clarification/addition.
   Don't mark -a as experimental anymore, and note that -a mode
   ("lazy", minimal-answers, w/o AUTH section in every reply) will
   be the default in future versions.  -A tells rbldnsd to go to
   "non-lazy" mode.  Document options in the manpage.

 - bugfix: fix configure script breakages:
   portability: for f; do => for f in "$@"; do
   fix broken GNU C (mis)detection

 - bugfix: fix dataset "un-expiration". Previously, once a dataset has expired,
   it never "un-expires" again even if new expire time is in future.  Due to
   missing reset of a dataset structure field.

 - portability: apparently at least one (broken) linux distribution includes
   kernel modification which leads to losing SIGALRM interrupts at times.
   So use setitimer() instead of alarm(), if available.

 - minor code cleanups here and there (fixing (real) GCC-4.1 warnings).

 - Debian-specific: adopt to more recent Debian packaging requiriments,
   and ensure that Debian package builds with zlib support

0.996 (19 Feb 2006)

 - experimental feature: data expiration support, in form
    $TIMESTAMP created expires
   see manpage for details.

 - feature: recognize new 'pass' entry in ACL "dataset", to allow
   whitelisting of a particular network/host covered by another
   ACL entry.

 - bugfix (sort of): deal with possible null-pointer dereferences on some
   systems such as FreeBSD where realloc(smaller_amount) may actually
   return NULL.  Note that this particular malloc implementation (where
   realloc() may return NULL if requested to reallocate to a smaller
   amount of memory) perform very badly with rbldnsd in the first place:
   rbldnsd tries to free some unused memory at the end of data load
   process, and realloc() forces a copy so there will be extra copy of

dns.h
/* common #include file for dns library.
 */

#ifndef DNS_PORT

#define DNS_PORT 53			/* default DNS port */
#define DNS_MAXPACKET 512		/* max size of UDP packet */
#define DNS_EDNS0_MAXPACKET 2048	/* max size of EDNS0 UDP packet */
#define DNS_MAXDN 255			/* max length of DN */
#define DNS_MAXLABEL 63			/* max length of one DN label */
#define DNS_MAXLABELS (DNS_MAXDN/2)	/* max # of labels in a DN */
#define DNS_MAXDOMAIN 1024		/* max length of asciiz DN */

enum dns_class {
  DNS_C_INVALID	= 0, /* invalid class */
  DNS_C_IN	= 1, /* Internet */
  DNS_C_CH	= 3, /* CHAOS */
  DNS_C_HS	= 4, /* HESIOD */
  DNS_C_ANY	= 255  /* wildcard */
};

enum dns_type {
  DNS_T_INVALID		= 0,	/* Cookie. */
  DNS_T_A		= 1,	/* Host address. */
  DNS_T_NS		= 2,	/* Authoritative server. */
  DNS_T_MD		= 3,	/* Mail destination. */
  DNS_T_MF		= 4,	/* Mail forwarder. */
  DNS_T_CNAME		= 5,	/* Canonical name. */
  DNS_T_SOA		= 6,	/* Start of authority zone. */
  DNS_T_MB		= 7,	/* Mailbox domain name. */
  DNS_T_MG		= 8,	/* Mail group member. */
  DNS_T_MR		= 9,	/* Mail rename name. */
  DNS_T_NULL		= 10,	/* Null resource record. */
  DNS_T_WKS		= 11,	/* Well known service. */
  DNS_T_PTR		= 12,	/* Domain name pointer. */
  DNS_T_HINFO		= 13,	/* Host information. */
  DNS_T_MINFO		= 14,	/* Mailbox information. */
  DNS_T_MX		= 15,	/* Mail routing information. */
  DNS_T_TXT		= 16,	/* Text strings. */
  DNS_T_RP		= 17,	/* Responsible person. */
  DNS_T_AFSDB		= 18,	/* AFS cell database. */
  DNS_T_X25		= 19,	/* X_25 calling address. */
  DNS_T_ISDN		= 20,	/* ISDN calling address. */
  DNS_T_RT		= 21,	/* Router. */
  DNS_T_NSAP		= 22,	/* NSAP address. */
  DNS_T_NSAP_PTR	= 23,	/* Reverse NSAP lookup (deprecated). */
  DNS_T_SIG		= 24,	/* Security signature. */
  DNS_T_KEY		= 25,	/* Security key. */
  DNS_T_PX		= 26,	/* X.400 mail mapping. */
  DNS_T_GPOS		= 27,	/* Geographical position (withdrawn). */
  DNS_T_AAAA		= 28,	/* Ip6 Address. */
  DNS_T_LOC		= 29,	/* Location Information. */
  DNS_T_NXT		= 30,	/* Next domain (security). */
  DNS_T_EID		= 31,	/* Endpoint identifier. */
  DNS_T_NIMLOC		= 32,	/* Nimrod Locator. */
  DNS_T_SRV		= 33,	/* Server Selection. */
  DNS_T_ATMA		= 34,	/* ATM Address */
  DNS_T_NAPTR		= 35,	/* Naming Authority PoinTeR */
  DNS_T_KX		= 36,	/* Key Exchange */
  DNS_T_CERT		= 37,	/* Certification record */
  DNS_T_A6		= 38,	/* IPv6 address (deprecates AAAA) */
  DNS_T_DNAME		= 39,	/* Non-terminal DNAME (for IPv6) */
  DNS_T_SINK		= 40,	/* Kitchen sink (experimentatl) */
  DNS_T_OPT		= 41,	/* EDNS0 option (meta-RR) */
  DNS_T_TSIG		= 250,	/* Transaction signature. */
  DNS_T_IXFR		= 251,	/* Incremental zone transfer. */
  DNS_T_AXFR		= 252,	/* Transfer zone of authority. */
  DNS_T_MAILB		= 253,	/* Transfer mailbox records. */
  DNS_T_MAILA		= 254,	/* Transfer mail agent records. */
  DNS_T_ANY		= 255,	/* Wildcard match. */
  DNS_T_ZXFR		= 256,	/* BIND-specific, nonstandard. */
  DNS_T_MAX		= 65536
};

enum dns_rcode {	/* reply code */
  DNS_R_NOERROR		= 0,	/* ok, no error */
  DNS_R_FORMERR		= 1,	/* format error */
  DNS_R_SERVFAIL	= 2,	/* server failed */
  DNS_R_NXDOMAIN	= 3,	/* domain does not exists */
  DNS_R_NOTIMPL		= 4,	/* not implemented */
  DNS_R_REFUSED		= 5,	/* query refused */
  /* these are for BIND_UPDATE */
  DNS_R_YXDOMAIN	= 6,	/* Name exists */
  DNS_R_YXRRSET		= 7,	/* RRset exists */
  DNS_R_NXRRSET		= 8,	/* RRset does not exist */
  DNS_R_NOTAUTH		= 9,	/* Not authoritative for zone */
  DNS_R_NOTZONE		= 10,	/* Zone of record different from zone section */
  /*ns_r_max = 11,*/
  /* The following are TSIG extended errors */
  DNS_R_BADSIG		= 16,
  DNS_R_BADKEY		= 17,
  DNS_R_BADTIME		= 18
};

struct dns_nameval {
  int val;
  const char *name;
};

extern const struct dns_nameval dns_classtab[];
extern const struct dns_nameval dns_typetab[];
extern const struct dns_nameval dns_rcodetab[];
const struct dns_nameval *
dns_findname(const struct dns_nameval *nv, const char *name);
#define dns_findclassname(class) dns_findname(dns_classtab, (class))
#define dns_findtypename(type) dns_findname(dns_typetab, (type))
#define dns_findrcodename(rcode) dns_findname(dns_rcodetab, (rcode))

const char *dns_classname(enum dns_class class);
const char *dns_typename(enum dns_type type);
const char *dns_rcodename(enum dns_rcode rcode);

unsigned dns_ptodn(const char *name, unsigned char *dn, unsigned dnsiz);
/* convert asciiz string `name' to the DN format, return length or 0 */
unsigned dns_dntop(const unsigned char *dn, char *dst, unsigned dstsiz);
unsigned dns_dntol(const unsigned char *srcdn, unsigned char *dstdn);
#define dns_dnlc(c) ((c) >= 'A' && (c) <= 'Z' ? (c) - 'A' + 'a' : (c))
unsigned dns_dnlen(const unsigned char *dn);
unsigned dns_dnlabels(const unsigned char *dn);
/* return number of labels in a dn */
int dns_dnequ(const unsigned char *dn1, const unsigned char *dn2);
unsigned dns_dnreverse(const unsigned char *dn, unsigned char *rdn,
                       unsigned len);
/* reverse order of labels in a dn; len, if non-zero, should be length
 * of dn (to save some CPU cycles) */

#endif
